File: queue/linkedlist_queue.py

from linked_list import ListNode, insert, remove, access, find, traverse

class LinkedListQueue:
	"""implement queue with linked list"""

	def __init__(self):
		"""constructor"""
		self._front: ListNode | None = None  # front node
		self._rear: ListNode | None = None  # rear node
		self._size: int = 0

	def size(self) -> int:
		"""get the length"""
		return self._size

	def is_empty(self) -> bool:
		"""check if it is empty"""
		return not self._front

	def push(self, num: int):
		"""enqueue"""
		# append element after rear node num
		node = ListNode(num)
		# if the queue is empty, both front and rear nodes point to the appended node
		if self._front is None:
			self._front = node
			self._rear = node
		# if the queue is not empty, the rear node point to the appended node
		else:
			self._rear.next = node
			self._rear = node
		self._size += 1

	def pop(self) -> int:
		"""dequeue"""
		num = self.peek()
		# delete front node
		self._front = self._front.next
		self._size -= 1
		return num
	
	def peek(self) -> int:
		"""get the front node"""
		if self.is_empty():
			raise IndexError("the queue is empty")
		return self._front.val

	def to_list(self) -> list[int]:
		"""convert queue to list"""
		queue = []
		temp = self._front
		while temp:
			queue.append(temp.val)
			temp = temp.next
		return queue

File: queue/linkedlist_queue.c
/* constructor */
LinkedListQueue *newLinkedListQueue() {
	LinkedListQueue *queue = (LinkedListQueue *)malloc(sizeof(LinkedListQueue));
	queue->front = NULL;
	queue->rear = NULL;
	queue->queSize = 0;
	return queue;
}

/* destructor */
void delLinkedListQueue(LinkedListQueue *queue) {
	// free all nodes
	for (int i = 0; i < queue->queSize && queue->front != NULL; i++) {
		ListNode *tmp = queue->front;
		queue->front = queue->front->next;
		free(tmp);
	}
	// free queue
	free(queue);
}

/* get queue size */
int size(LinkedListQueue *queue) {
	return queue->queSize;
}

/* check if queue is empty */
bool empty(LinkedListQueue *queue) {
	return (size(queue) == 0);
}

/* enqueue */
void push(LinkedListQueue *queue, int num) {
	// append a new node to the end of queue
	ListNode *node = newListNode(num);
	// if queue is empty, set front and rear to the new node
	if (queue->front == NULL) {
		queue->front = node;
		queue->rear = node;
	}
	// if queue is not empty, append the new node to the end of queue
	else {
		queue->rear->next = node;
		queue->rear = node;
	}
	queue->queSize++;
}

/* get the front element */
int peek(LinkedListQueue *queue) {
	assert(size(queue) && queue->front);
	return queue->front->val;
}

/* dequeue */
int pop(LinkedListQueue *queue) {
	int num = peek(queue);
	ListNode *tmp = queue->front;
	queue->front = queue->front->next;
	free(tmp);
	queue->queSize--;
	return num;
}

/* print queue */
void printLinkedListQueue(LinkedListQueue *queue) {
	int *arr = malloc(sizeof(int) * queue->queSize);
	// copy queue to array
	int i;
	ListNode *node;
	for (i = 0, node = queue->front; i < queue->queSize; i++) {
		arr[i] = node->val;
		node = node->next;
	}
	printArray(arr, queue->queSize);
	free(arr);
}